lib/core/src/irods_default_paths.cpp

```diff
@@ -18,31 +21,83 @@ namespace irods {
         try {
             boost::filesystem::path path{dl_info.dli_fname};
             path = boost::filesystem::canonical(path);
-            path.remove_filename().remove_filename().remove_filename(); // Removes filename and the two directories (usr and lib) between libirods_common.so and base of irods install
+            path.remove_filename();
             return path;
         } catch(const boost::filesystem::filesystem_error& e) {
             THROW(-1, e.what());
         }
     }
 
+    boost::filesystem::path
+    get_irods_root_directory() {
+        try {
+            boost::filesystem::path path = get_irods_lib_directory();
+            boost::filesystem::path install_libdir{IRODS_DEFAULT_PATH_LIBDIR};
+            install_libdir = install_libdir.lexically_normal();
+            long install_path_len = std::distance(install_libdir.begin(), install_libdir.end());
+            if (install_libdir.is_absolute()) {
+                install_path_len--;
+            }
+            for (long i = 0; i < install_path_len; i++) {
+                path.remove_filename(); // Removes directories (usually usr and lib) between libirods_common.so and base of irods install
+            }
+            return path;
+        } catch(const boost::filesystem::filesystem_error& e) {
+            THROW(-1, e.what());
```

Comment:
It may be worth using a better error code than -1. You can check `rodsErrorTable.h` to see if there is something more descriptive. If no existing error code works for this case, you can add a new one.

Reply:
`SYS_INTERNAL_ERR` sound good? Probably as specific as we can get without dissecting the caught exception.

Reply:
Sure. That will at least allow the client/server to resolve the error code to an error name. I don't think -1 maps to any existing error code/name at this time.